### ISSUE TITLE
valve setting parameter not used as expected

### ISSUE THREAD

**Issue by Nathan-Becker** [glorialulu/wmoc#8]
(I posted this earlier, but I think it was mixed up with other issues. Reposting separately.)

The setting parameter of the FCV valve doesn't seem to be used by the transient simulation. I was expecting that if I set a FCV valve to a particular setting then the flow-rate should match it.

The following test code might be revealing. There are two parameter settings at the top: valve_setting_gpm and initial_status_active. I looked at 2 cases:

If I set initial_status_active=True, then I think that is consistent with the EPANET manual's claim that the valve is allowed to operate via control messages. In that case, if I run the below script to the end, it prints out the initial flow-rate, and it is consistent with the valve_settting_gpm value. However, the final flow-rate (after the valve opens and reaches steady state) is not consistent with the valve_setting_gpm. And, there is some noisy ringing at the beginning of the transient simulation (shown in the plots that are generated). Maybe this ringing is caused by some sort of inconsistency between EPANET valve initial condition (which seems to respect the valve_setting parameter) and the transient simulator?
If I set initial_status_active=False, then the valve is forced closed at the beginning. The noisy ringing goes away, and the plots look qualitatively plausible. However, the valve_setting_gpm seems to have no effect on the steady-state flow-rate. I expected the steady-state flow-rate to match the valve_setting_gpm if the pump can support it.

```
# Define test-case options
valve_setting_gpm = 2.56
initial_status_active = True

# Initialize
import matplotlib
matplotlib.use('agg')
from matplotlib import pyplot as plt
import numpy
import pandas
import wntr
import tsnet


def gpm_to_m3s(gpm):
    return gpm/(60*264.172)

def m3s_to_gpm(m3s):
    return m3s*(60*264.172)

def m_to_psi(m):
    return m*1.4219702063247

def psi_to_m(psi):
    return psi/1.4219702063247

# New network
wn = wntr.network.WaterNetworkModel()

# Add water source --> reservoir at atmos, with a pump to get significant pressure
wn.add_reservoir('res', base_head=0)
wn.add_junction('in')
wn.add_pipe('source', start_node_name='res', end_node_name='in', length=40, diameter=0.03175)
wn.add_junction('pumpjunc')
pumpcurve = []
pumpcurve.append((gpm_to_m3s(0),psi_to_m(60)))
pumpcurve.append((gpm_to_m3s(5),psi_to_m(45)))
pumpcurve.append((gpm_to_m3s(10),psi_to_m(25)))
wn.add_curve('pumpcurve', 'HEAD', pumpcurve)
wn.add_pump(name='supply', start_node_name='in', end_node_name='pumpjunc', pump_type='HEAD', pump_parameter='pumpcurve')

# Add reference point junction 'ref'
wn.add_junction('ref')
wn.add_pipe('to_ref', start_node_name='pumpjunc', end_node_name='ref', length=10, diameter=0.03175)

# Add valve
wn.add_junction('before_valve1')
wn.add_pipe('to_valve1', start_node_name='ref', end_node_name='before_valve1', length=25, diameter=0.03175)
wn.add_junction('after_valve1')
wn.add_valve('valve1', start_node_name='before_valve1', end_node_name='after_valve1', valve_type='FCV', setting=gpm_to_m3s(valve_setting_gpm))
if initial_status_active:
    wn.get_link('valve1').initial_status = wntr.network.elements.LinkStatus.Active
else:
    wn.get_link('valve1').initial_status = wntr.network.elements.LinkStatus.Closed
wn.add_reservoir('atmos1', base_head=psi_to_m(0))
wn.add_pipe('out1', start_node_name='after_valve1', end_node_name='atmos1', length=15, diameter=0.03175/3)

# Prep plotting
fig = plt.figure(figsize=(8,5), dpi=80, facecolor='w', edgecolor='k')

# Build EPANET file
wn.write_inpfile('simp.inp')

# Construct transient model 
tm = tsnet.network.TransientModel('simp.inp')
tm.set_wavespeed(1200.) # m/s
tm.set_time(3) # simulation period [s]

# Set valve opening 
tc = 0.2 # valve opening period [s]
ts = 1 # valve opening start time [s]
se = 1 # end open percentage [s]
m = 1 # open constant [dimensionless]
tm.valve_opening('valve1',[tc,ts,se,m])

# Initialize and run simulation
tm = tsnet.simulation.Initializer(tm, 0, 'DD')
tm = tsnet.simulation.MOCSimulator(tm, 'simpres', 'unsteady')

# Gather results
t = tm.simulation_timestamps
ref_flow_rate = tm.get_link('to_ref').end_node_flowrate

# Prep plotting
fig = plt.figure(figsize=(8,5), dpi=80, facecolor='w', edgecolor='k')

# Plot flow-rates
plt.plot(t, m3s_to_gpm(ref_flow_rate), label='ref', linewidth=1.5)
plt.plot(t, m3s_to_gpm(tm.get_link('out1').end_node_flowrate), label='output1', linewidth=1.5)
plt.ylabel("flow-rate [gpm]")
plt.xlim([t[0],t[-1]])
plt.xlabel("Time [s]")
plt.legend(loc='best')
plt.draw()
plt.savefig('flowrate.png')

# Plot pressure
plt.clf()
plt.plot(t, m_to_psi(tm.get_node('ref').head), 'k', label='ref', linewidth=1.5)
plt.plot(t, m_to_psi(tm.get_node('after_valve1').head), label='output1', linewidth=1.5)
plt.ylabel("Pressure [PSI]")
plt.xlim([t[0],t[-1]])
plt.xlabel("Time [s]")
plt.legend(loc='best')
plt.draw()
plt.savefig('pressure.png')

print('Initial ref flow rate={}'.format(m3s_to_gpm(ref_flow_rate[0])))
print('Final ref flow rate={}'.format(m3s_to_gpm(ref_flow_rate[-1])))
```

**Comment by glorialulu** [comment]
Hi Nathan, 

The operating valves in TSNet are controlled by the loss coefficient (ls), which is specified as a function of opening percentage (p), i.e., ls = f(p). This function is not used in the steady-state simulation but required for transient simulation if the valve is to be operated. It can be defined using:
`valve_opening(name, rule, curve=None)`, as shown in the [API documentation](https://tsnet.readthedocs.io/en/latest/apidocs/tsnet.network.html?highlight=valve_opening#tsnet.network.model.TransientModel.valve_opening)
A default valve curve is included in TSNet, but a customized curve for the specific valve can and should be supplied.

The opening percentage is then specified the valve operation curve: p = g(t).

By the end, combing ls=f(p) and p =g(t), the valve operation is controlled by the changing loss coefficient. 
Thus, it is not designed for FCV, where flowrate is the control variable.

**Comment by glorialulu** [comment]
The noisy ringing at the beginning of the transient simulation is indeed by the inconsistency between EPANET valve initial condition and the transient simulator, because disagreement exists between the loss coefficient before opening ls(p=0) as specified by the valve curve and the setting in EPANet when the initial status is active.

Thus, my suggestion for your problem is to find the appropriate valve curve.

**Comment by Nathan-Becker** [comment]
Thanks for the additional information. Could you please give some further details? I experimented with several different curves, but I'm not getting outputs that I expect.

Here's what I tried:

In the above script, I set initial_status_active = False. Then I changed the valve opening line to:
```
tm.valve_opening('valve1',[tc,ts,se,m], curve=[(i, 1e-3) for i in numpy.linspace(0,1,11)])
```
In that case, I do find that the flow reaches steady-state around 1.2 gpm. But, the valve apparently opens immediately, at t=0s, rather than waiting until t=1s as specified.
I tried another case, where the loss follows a non-constant behavior:
```
tm.valve_opening('valve1',[tc,ts,se,m], curve=[(i, 1e-3*(i+1e-3)/(1+1e-3)) for i in numpy.linspace(0,1,11)])
```
This curve gives a totally different behavior, despite having the same loss coefficient when the valve is fully-open at p=1. In this case, I find the flow reaches steady state of about 1.2 gpm during the first second (when the valve was supposed to be closed.) Then at t=1s, the flow suddenly drops to zero and rings a bit during the remaining time of the simulation.

**Comment by glorialulu** [comment]
Hi, 

The first valve curve you experiment is a constant-loss valve, which means that no matter how and when you close/open the valve, the valve coefficient will not change. Also, you made that constant to be a fairly small number, which basically says that there is little loss at the valve no matter how you operate it. However, in the initial condition calculated by EPANet, there is loss as the valve. This is why you see an immediate opening, and then the flows recover back to normal, 1.2gpm.

The second valve curve has an increasing loss coefficient as the valve opens, which is not physical. In general, the higher the loss coefficient, the slower the flow.

In fact, the results will start to make more sense when I reverse the second valve curve:
`tm.valve_opening('valve1',[tc,ts,se,m], curve=[(i, 1e-3*(i+1e-3)/(1+1e-3)) for i in numpy.linspace(100,0,11)])`

![image](https://user-images.githubusercontent.com/32466099/89319221-6f807b00-d645-11ea-82bd-208ad31c6c82.png)

There are still some ringing near the beginning, which is due to the inconsistency of valve loss in EPANet and the ls(0) that you specified. 

The default valve curve defined in [TSNet](https://github.com/glorialulu/TSNet/blob/master/tsnet/utils/valve_curve.py) looks like this:
![image](https://user-images.githubusercontent.com/32466099/89318170-0c421900-d644-11ea-92fb-d2960585b7fe.png) 

The loss coefficient is near infinite when the valve is fully closed, and a small value when it is fully open, which equals the head loss in the EPANet model.

**Comment by Nathan-Becker** [comment]
Thanks very much for the additional details and for your patience in continuing to answer my questions.

I experimented with what you said, but I'm still a bit confused about some things. I wrote up a new version of the test program to work through my questions:

1. I played around with setting 'minor_loss' and 'setting' for the valve. I found that I could diminish the ringing at the beginning of the simulation by setting minor_loss=1e10, for setting=1. However, this doesn't seem like the cleanest way to control things, so I switched back to minor_loss=0 and setting=0, which completely eliminates the ringing, presumably because the initial conditions in EPANET are for flow_rate=0 due to the setting=0 value.

2. I experimented with alternate valve curves. I see what you are saying about the curve I tried previously being non-physical. However, I'm confused by the apparent default curve, which also seems non-physical: when I plot it, I don't seem to get the same curve that you showed in your previous comment. This is my plot (generated by code below):
![valve_curve](https://user-images.githubusercontent.com/53197475/89356168-f9463d80-d671-11ea-8eae-06b0937ceafb.png)

The program below investigates different valve curves. It uses the same curve-shape as the default, but just multiplies the loss by a constant kl_fact, defined at the top. For a value of kl_fact=1, it also does some consistency checks.

I find that the steady-state flow-rate gets smaller as the loss is reduced, which is counter to what I was expecting. For kl_fact=1, the steady-state is around 6gpm. For kl_fact=0.001 the steady state is around 2.5 gpm.

Also, I'm confused on what the units of "percent open" should be. Is it supposed to range from 0 to 100 or 0 to 1? From the code here(https://github.com/glorialulu/TSNet/blob/4c3aa09585a0c28d6be86bb37d0007e731045569/tsnet/utils/valve_curve.py#L24), it seems like it should range from 0 to 100. But, from the use of valve_curve in the simulation (for example, here: https://github.com/glorialulu/TSNet/blob/4c3aa09585a0c28d6be86bb37d0007e731045569/tsnet/simulation/main.py#L161), it seems that "open percentage" should range from 0 to 1 because it is then multiplied by 100. I wonder if this is causing a bug where the valve is effectively jumping from completely closed to completely open because the open percent rapidly reaches 100, causing the interpolation in valve_curve to return the max value for most "open percent" values passed by the simulation. Perhaps such a rapid opening of the valve would explain the large amount of ringing I see when setting provide_curve=False in the test code below, which then uses the default valve curve.

In summary, could you please answer these questions:
a) What is the expected range for "open percentage" (both in defining the valve curve and in the tm.valve_opening(...) call)
b) Why is the steady-state flow-rate apparently lower for smaller kl_fact?
c) Is there a more appropriate way to configure minor_loss and setting so that the initial conditions of EPANET match the transient sim?

thanks again!

```
# Define test-case options
initial_status_active = True
provide_curve = True
kl_fact = 1
#kl_fact = 0.001
minor_loss = 0
setting = 0
#minor_loss = 1e10
#setting = 1

# Initialize
import matplotlib
matplotlib.use('agg')
from matplotlib import pyplot as plt
import numpy
import pandas
import wntr
import tsnet
from valve_curve import valve_curve


def gpm_to_m3s(gpm):
    return gpm/(60*264.172)

def m3s_to_gpm(m3s):
    return m3s*(60*264.172)

def m_to_psi(m):
    return m*1.4219702063247

def psi_to_m(psi):
    return psi/1.4219702063247

# New network
wn = wntr.network.WaterNetworkModel()

# Add water source --> reservoir at atmos, with a pump to get significant pressure
wn.add_reservoir('res', base_head=0)
wn.add_junction('in')
wn.add_pipe('source', start_node_name='res', end_node_name='in', length=40, diameter=0.03175)
wn.add_junction('pumpjunc')
pumpcurve = []
pumpcurve.append((gpm_to_m3s(0),psi_to_m(60)))
pumpcurve.append((gpm_to_m3s(5),psi_to_m(45)))
pumpcurve.append((gpm_to_m3s(10),psi_to_m(25)))
wn.add_curve('pumpcurve', 'HEAD', pumpcurve)
wn.add_pump(name='supply', start_node_name='in', end_node_name='pumpjunc', pump_type='HEAD', pump_parameter='pumpcurve')

# Add reference point junction 'ref'
wn.add_junction('ref')
wn.add_pipe('to_ref', start_node_name='pumpjunc', end_node_name='ref', length=10, diameter=0.03175)

# Add valve
wn.add_junction('before_valve1')
wn.add_pipe('to_valve1', start_node_name='ref', end_node_name='before_valve1', length=25, diameter=0.03175)
wn.add_junction('after_valve1')
wn.add_valve('valve1', start_node_name='before_valve1', end_node_name='after_valve1', valve_type='FCV', minor_loss=minor_loss, setting=setting)
if initial_status_active:
    wn.get_link('valve1').initial_status = wntr.network.elements.LinkStatus.Active
else:
    wn.get_link('valve1').initial_status = wntr.network.elements.LinkStatus.Closed
wn.add_reservoir('atmos1', base_head=psi_to_m(0))
wn.add_pipe('out1', start_node_name='after_valve1', end_node_name='atmos1', length=15, diameter=0.03175/3)

# Prep plotting
fig = plt.figure(figsize=(8,5), dpi=80, facecolor='w', edgecolor='k')

# Build EPANET file
wn.write_inpfile('simp.inp')

# Construct transient model 
tm = tsnet.network.TransientModel('simp.inp')
tm.set_wavespeed(1200.) # m/s
tm.set_time(3) # simulation period [s]

# Extract default valve curve
po = numpy.linspace(100,0,11)
kl = valve_curve(po)

# Scale loss by requested factor and assemble the curve in the required format
kl = kl*kl_fact
curve = list(zip(po, kl))

# Confirm 'curve' contains the default curve
if kl_fact == 1:
    curve_test1 = valve_curve(po, coeff=(po, kl))
    curve_test2 = valve_curve(po, coeff=list(zip(*curve)))
    assert((curve_test1 == kl).all())
    assert((curve_test2 == kl).all())
    print('default valve_curve passed consistency check')

# Plot the valve curve
plt.clf()
plt.plot(po, kl, 'o')
plt.xlabel('% open')
plt.ylabel('loss')
plt.draw()
plt.savefig('valve_curve.png')

# Set valve opening 
tc = 0.2 # valve opening period [s]
ts = 1 # valve opening start time [s]
se = 1 # end open percentage [s]
m = 1 # open constant [dimensionless]
if provide_curve:
    tm.valve_opening('valve1',[tc,ts,se,m], curve=curve)
else:
    tm.valve_opening('valve1',[tc,ts,se,m])

# Initialize and run simulation
tm = tsnet.simulation.Initializer(tm, 0, 'DD')
tm = tsnet.simulation.MOCSimulator(tm, 'simpres', 'unsteady')

# Gather results
t = tm.simulation_timestamps
ref_flow_rate = tm.get_link('to_ref').end_node_flowrate

# Prep plotting
fig = plt.figure(figsize=(8,5), dpi=80, facecolor='w', edgecolor='k')

# Plot flow-rates
plt.clf()
plt.plot(t, m3s_to_gpm(ref_flow_rate), label='ref', linewidth=1.5)
plt.plot(t, m3s_to_gpm(tm.get_link('out1').end_node_flowrate), label='output1', linewidth=1.5)
plt.ylabel("flow-rate [gpm]")
plt.xlim([t[0],t[-1]])
plt.xlabel("Time [s]")
plt.legend(loc='best')
plt.draw()
plt.savefig('flowrate.png')

# Plot pressure
plt.clf()
plt.plot(t, m_to_psi(tm.get_node('ref').head), 'k', label='ref', linewidth=1.5)
plt.plot(t, m_to_psi(tm.get_node('after_valve1').head), label='output1', linewidth=1.5)
plt.ylabel("Pressure [PSI]")
plt.xlim([t[0],t[-1]])
plt.xlabel("Time [s]")
plt.legend(loc='best')
plt.draw()
plt.savefig('pressure.png')

print('Initial ref flow rate={}'.format(m3s_to_gpm(ref_flow_rate[0])))
print('Final ref flow rate={}'.format(m3s_to_gpm(ref_flow_rate[-1])))
```

**Comment by glorialulu** [comment]
Hi, 

To your questions: 

(a) valve_curve.py take open_percentage from 0 to 100. tm.valve_opening take input as ranging from 0 to 1. This is why you see  `valve[0] = valve_curve(tm.links[utype[pn][1]].operation_rule[ts]*100, ` in main.py.

The reason why you see an increasing valve curve is that the kl is defined as 1/loss coefficient. The reason why I chose this implementation is that when open_percentage = 0, the loss coefficient is infinite, while 1/loss coefficient is finite and equals 0.

Thus, I take back what I said in my last response. The loss coefficient should decrease as open_percentage changes from 0 to 100, but kl=1/loss_coefficient should increase.

b) Why is the steady-state flow-rate apparently lower for smaller kl_fact?

As I said in response to your previous question. loss coefficient = 1/kl. So when kl_fact is smaller, the valve loss coefficient is larger, which induces more head loss and leads to a lower flow rate.

c) Is there a more appropriate way to configure minor_loss and setting so that the initial conditions of EPANET match the transient sim?

The loss coefficient for an initially active FCV is kind of tricky, because TSNet kind of assume the initial status of a valve is either open or close.  However, the open percentage of FCV is dependent on the flow conditions. I will need to dig deeper to find more elegant solutions. If you can set the initial FCV status as inactive, the setting is much more straightforward.

**Comment by Nathan-Becker** [comment]
Thank you, I think this makes sense to me now.
